src/python/pants/java/nailgun_client.py

```diff
@@ -57,6 +58,12 @@ def _process_session(self):
           self._maybe_start_input_reader()
         else:
           raise self.ProtocolError('received unexpected chunk {} -> {}'.format(chunk_type, payload))
+    except (IOError, OSError) as e:
+      # If a `Broken Pipe` is encountered during a stdio fd write, we're headless - bail.
```

Comment:
I'm guessing that you should only get `EPIPE` when dealing with `_stdout`/`_stderr` here, and never when dealing with the socket?